### PR TITLE
Parse <collision_default> and <enable_collisions> tags

### PR DESCRIPTION
**Pull request by rhaschke**
This PR augments the SRDF XML with tags `<collision_default>` and `<enable_collisions>` to allow specifying ACM defaults for individual links. For example,
```xml
<collision_default link="panda_hand_coarse" allow="ALWAYS" />
<enable_collisions link1="panda_hand_coarse" link2="panda_link0" />
```
UPDATE: See [below](https://github.com/ros-planning/srdfdom/pull/97#pullrequestreview-838696808) for the final xml tags.

will disable collision checking for `panda_hand_coarse` with all other links (particularly newly introduced objects) but enable collision between `panda_hand_coarse` and `panda_link0`.

### CODE REVIEW

File: src/model.cpp
Comment (v4hn) [inline src/model.cpp]:
```suggestion
      CONSOLE_BRIDGE_logWarn("Link '%s' is not known to URDF. Cannot specify collision default.", link_);
```

File: src/model.cpp
Comment (v4hn) [inline src/model.cpp]: You only permit a single value here and it does not affect anything. Do you have alternatives in mind for the future?
I think `<collision_default allow="ALWAYS">` is not very readable. How about a simpler tag with the same semantics
```
<disable_default_collisions link=""/>
```
or alternatively a wildcard in the existing tag
```
<disable_collisions link1="tool0" link2="*"/>
```
?

File: include/srdfdom/model.h
Comment (v4hn) [inline include/srdfdom/model.h]:
```suggestion
  const std::vector<std::string>& getNoDefaultCollisionLinks() const
```
... and of course the associated refactoring.

File: src/srdf_writer.cpp
Comment (v4hn) [inline src/srdf_writer.cpp]:
```suggestion
  for (const srdf::Model::CollisionPair& pair : collision_pairs_)
```

File: src/srdf_writer.cpp
Comment (v4hn) [inline src/srdf_writer.cpp]: It seems very weird to call this method here instead of at the end of `createCollisionPairsXML` below.

File: include/srdfdom/model.h
Comment (v4hn) [inline include/srdfdom/model.h]:
```suggestion
  std::vector<CollisionPair> disabled_collision_pairs_;
  std::vector<CollisionPair> enabled_collision_pairs_;
```
I believe this would make the code below somewhat nicer and you could drop the `disabled_` flag in each object.

Feel free to disregard this comment if you disagree.

**Review by v4hn — COMMENTED** [review]
Thanks for working on this. It's quite some work for a (at the moment) edge-case feature :+1:

File: src/srdf_writer.cpp
Comment (rhaschke) [reply src/srdf_writer.cpp]: Logically, re-enabling collisions is linked to the function `createCollisionDefaultsXML()` here, isn't it?
The method `createCollisionPairsXML(XMLElement*)` refers to the old method called to disable collision pairs.

File: include/srdfdom/model.h
Comment (rhaschke) [reply include/srdfdom/model.h]: If you split this information into two separate members, you need to duplicate the corresponding code to handle them, don't you? Where exactly do you see a simplification?

File: src/srdf_writer.cpp
Comment (v4hn) [reply src/srdf_writer.cpp]: No hard requirement from my side to pass review. I would just not expect something called "create collision defaults" to generate collision **pairs** as well.
I would expect that to be done by calling "create collision pairs" instead :cat2: 
Otherwise you might want to consider renaming methods to reflect this better.

File: include/srdfdom/model.h
Comment (v4hn) [reply include/srdfdom/model.h]: I guess I mostly dislike [passing a bool to a function just to filter a list](https://github.com/ros-planning/srdfdom/pull/97/files#diff-e7f7ff52129fe82e20745eef9247c07c322e01e63aa6debc6edad556b7876ca3R353). You might just as well pass in the separate lists as arguments and avoid iterating over data you know you'll skip...

Again, I don't feel strongly about it.

Comment (rhaschke) [reply include/srdfdom/model.h]: I see your point. I have split the vectors now.

File: src/srdf_writer.cpp
Comment (rhaschke) [reply src/srdf_writer.cpp]: I want to keep it here.

**Review by JafarAbdi — APPROVED** [review]
Thank you for tackling this, is there something blocking merging this PR?

The example in the PR description is outdated, right? it should be something like

```xml
<disable_default_collisions link="link_name" />
<enable_collisions link1="link1_name" link2="link2_name" reason="optional-reason" />
<disable_collisions link1="link1_name" link2="link2_name" reason="optional-reason" />
```

**Comment by rhaschke** [comment]
Yes, as Michael asked for an adaption of the XML tags, the example should be adapted correspondingly.
This PR should be merged in sync with https://github.com/ros-planning/moveit/pull/2938. However, nobody reviewed that PR yet.

**Comment by v4hn** [comment]
> This PR should be merged in sync with ros-planning/moveit#2938.

Why in sync? I was about to merge this now as the basis for the MoveIt patch. Feel free to merge this now if you agree.

> However, nobody reviewed that PR yet.

Sorry, I did not have enough time for both yet, so I focused on the basis first.

**Comment by rhaschke** [comment]
> Why in sync?

This PR changes API and thus breaks MoveIt. Thus it should be merged in sync with https://github.com/ros-planning/moveit/pull/2938, which is currently broken due to your requested API changes.

**Comment by v4hn** [comment]
> This PR changes API and thus breaks MoveIt.

Of course. ignore me, I have no idea what I was thinking...

> Thus it should be merged in sync with https://github.com/ros-planning/moveit/pull/2938

:+1:

> which is currently broken due to your requested API changes.

I will not manage before Christmas, but I could review after and once CI succeeds again.
But maybe @JafarAbdi is faster? ;)

**Comment by kbrameld** [comment]
I'd like to ask for some clarification in regards to these changes.

If a robot has a link that has collision disabled by default (via ``disable_default_collisions``), does that mean that link does not collide with other collision objects in the environment (eg. table, wall) and other robots? Or does it just disable collisions between links in the robot.

**Comment by rhaschke** [comment]
It means disabling collisions with any robot link or world object.

**Comment by kbrameld** [comment]
Thanks @rhaschke

**Pull request merged**
